cocos/core/math/mat4.ts

```diff
@@ -930,27 +900,43 @@ export class Mat4 extends ValueType {
         const sx = Vec3.set(v3_1, m.m00, m.m01, m.m02).length();
         const sy = Vec3.set(v3_1, m.m04, m.m05, m.m06).length();
         const sz = Vec3.set(v3_1, m.m08, m.m09, m.m10).length();
-        if (s) {
-            s.x = sx;
-            s.y = sy;
-            s.z = sz;
-        }
-        if (v) {
-            Vec3.set(v, m.m12, m.m13, m.m14);
-        }
-        if (q) {
+
+        if (sx !== 0) {
             m3_1.m00 = m.m00 / sx;
             m3_1.m01 = m.m01 / sx;
             m3_1.m02 = m.m02 / sx;
+        } else {
+            m3_1.m00 = m3_1.m01 = m3_1.m02 = 0;
+        }
+        if (sy !== 0) {
             m3_1.m03 = m.m04 / sy;
             m3_1.m04 = m.m05 / sy;
             m3_1.m05 = m.m06 / sy;
+        } else {
+            m3_1.m03 = m3_1.m04 = m3_1.m05 = 0;
+        }
+        if (sz !== 0) {
             m3_1.m06 = m.m08 / sz;
             m3_1.m07 = m.m09 / sz;
             m3_1.m08 = m.m10 / sz;
-            const det = Mat3.determinant(m3_1);
+        } else {
+            m3_1.m06 = m3_1.m07 = m3_1.m08 = 0;
+        }
+
+        const det = Mat3.determinant(m3_1);
+        if (s) {
+            s.x = sx;
+            s.y = sy;
+            s.z = sz;
+            if (det < 0) {
+                s.x *= -1;
+            }
+        }
```

Comment:
Here I don't understand, why inverse s.x?

Reply:
https://github.com/cocos/cocos-engine/blob/v3.8.3/cocos/core/math/mat4.ts#L904

It's from the original logic。

### There is an explain from AI:


When the determinant is less than 0, it signifies two important physical meanings:

Reversal of Orientation: The sign of the determinant indicates whether a transformation preserves the orientation of space. When the determinant is negative, it means that the transformation reverses the orientation of space, flipping from the positive direction to the negative direction. Geometrically, this implies that objects are oriented oppositely after the transformation compared to the original orientation.
Volume Reversal: The absolute value of the determinant represents the scaling factor of the space volume defined by the transformation. When the determinant is negative, it implies that the transformation results in a reversal of the space volume, meaning that after the transformation, the volume of objects in space is reversed, becoming opposite.

Reply:
@star-e I found some reference in Ogre engine:

https://github.com/OGRECave/ogre/blob/26c02c910dcc3c90580111403e6e40285ee785ad/OgreMain/src/OgreMatrix3.cpp#L626

Reply:
I think it mirrors the x-axis. As long as the conversion is consistent, it will be fine.